rfcs/convergence/root-as-slot.md

```diff
@@ -0,0 +1,248 @@
+  // ...
+  return {
+    ...props,
+    slot: resolveShorthand(slot),
+  };
+}
+```
+
+With the modifications:
+
+```ts
+export type Slots = {
+  root: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>, HTMLElement>;
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> {
+  /* ... */
+}
+export interface State extends ComponentState<Slots> {
+  /* ... */
+}
+
+// use*State hook
+function useState({slot, ...props}: Props): State {
+  // ...
+  return {
+    slot: resolveShorthand(slot)
+    root: getNativeElementProps(state.components.root, props),
+  };
+}
+```
+
+### Make _slots_ support `as` property (but with restrictions)
+
+Cancels [Do not support as prop for slots that render native DOM elements] as it already conflicts with [1st rule of ARIA opt-out mechanism]. This would add `as` property for `ObjectShorthandProps` signature for all _slots_.
+
+To ensure `as` property is used for opt-out mechanism only native elements will be supported. _Typings_ must be used to ensure which native elements are supported.
+
+`ObjectShorthandProps` should be a model that _shorthand_ properties should extend from, and to ensure `as` property is only provided in cases that opt-out mechanism is required the `never` type should be used to disable `as` property on default cases.
+
+```ts
+type ObjectShorthandProps<Props = {}, Ref = unknown, As extends keyof JSX.IntrinsicElements = never> = Props &
+  React.RefAttributes<Ref> & {
+    as?: As;
+    children?: Props['children'] | ShorthandRenderFunction<Props>;
+  };
+```
+
+The case for `useARIAButton` is a good example of using this signature:
+
+```ts
+export type ARIAButtonShorthandProps =
+  | ObjectShorthandProps<JSX.IntrinsicElements['button'], HTMLButtonElement, /*as*/ 'button'>
+  | ObjectShorthandProps<JSX.IntrinsicElements['div'], HTMLDivElement, /*as*/ 'div'>
+  | ObjectShorthandProps<JSX.IntrinsicElements['span'], HTMLSpanElement, /*as*/ 'span'>
+  | ObjectShorthandProps<JSX.IntrinsicElements['a'], HTMLAnchorElement, /*as*/ 'a'>;
+```
+
+`ARIAButtonProps` is a union between types from `button`, `div`, `span` and `anchor` native elements.
+
+> ⚠️ using union types for declaring properties is not perfect https://catchts.com/unions
+
+### No reason for `getNativeElementProps`
+
+The problem reported on [Widening Types](#⚠️-Widening-types-problem-⚠️), makes this a little bit controversial.
+
+Although types should be enough to ensure that only supported properties will be passed to a _slot_, [Widening Types problem](#⚠️-Widening-types-problem-⚠️) make it not so much true for `root` at least. Besides `root` is safe to say that `getNativeElementProps` is not necessary.
+
+### Add references to all _slots_
+
+Right now _ref_ is not supported on _Typings_ for _slots_. This proposes that _ref_ should be part of the internal interface of `ObjectShorthandProps` to allow access of references to internal slots of a given component.
```

Comment:
Nit, I think it would be good to use code formatting (backticks) for things like `ref` that match up to actual names in code.
```suggestion
Right now `ref` is not supported on _Typings_ for _slots_. This proposes that `ref` should be part of the internal interface of `ObjectShorthandProps` to allow access of references to internal slots of a given component.
```